addon/components/gavant-table.ts

```diff
@@ -0,0 +1,532 @@
+        const newColumnPanPosition = this.columnPanPosition + moveIndex;
+        const pannedColumns = this.allowFixedCols ? this.nonFixedColumns : this.columns;
+        if (newColumnPanPosition < 0 || newColumnPanPosition >= pannedColumns.length) {
+            return;
+        }
+
+        set(this, 'columnPanPosition', newColumnPanPosition);
+        this.updateColumnVisibility();
+    }
+
+    /**
+     * Returns the computed width for the given element
+     *
+     * @param {Element} el -- A DOM element
+     * @returns {number}
+     * @memberof GavantTableComponent
+     */
+    getElementWidth(el: Element): number {
+        let width = 0;
+
+        if (el) {
+            let rects = el.getClientRects();
+            let paddingLeft = parseFloat(window.getComputedStyle(el, null).getPropertyValue('padding-left'));
+            let paddingRight = parseFloat(window.getComputedStyle(el, null).getPropertyValue('padding-right'));
+            if (rects) {
+                width = paddingLeft && paddingRight ? rects[0].width - (paddingLeft + paddingRight) : rects[0].width;
+            }
+        }
+
+        return width;
+    }
+
+    /**
+     * Returns the column's unique identifier or undefined
+     *
+     * @param {(TableColumn | undefined)} col - A table column
+     * @returns {(string | undefined)}
+     * @memberof GavantTableComponent
+     */
+    getColumnId(col: TableColumn | undefined): string | undefined {
+        //allow columns to specify an "id" if they dont have a valuePath, or the valuePath is not unique
+        return col && (col.id || col.valuePath);
+    }
+
+    //actions
+
+    /**
+     * Invokes an action to load a new page of rows
+     *
+     * @returns {Promise}
+     * @memberof GavantTableComponent
+     */
+    @action
+    onLastReached() {
+        if (!this.isLoading && this.hasMoreRows) {
+            return tryInvoke(this, 'loadMoreRows');
+        }
+        return;
+    }
+
+    /**
+     * Pans the table's visible columns to the left
+     *
+     * @returns {void}
+     * @memberof GavantTableComponent
+     */
+    @action
+    panColumnsLeft() {
+        return this.panColumns(-1);
+    }
+
+    /**
+     * Pans the table's visible columns to the right
+     *
+     * @returns {void}
+     * @memberof GavantTableComponent
+     */
+    @action
+    panColumnsRight() {
+        return this.panColumns(1);
+    }
+
+    /**
+     * Updates the table's column sorting
+     *
+     * @param {any[]} sorts
+     * @returns {Promise}
+     * @memberof GavantTableComponent
+     */
+    @action
+    onUpdateSorts(sorts: any[]) {
+        set(this, 'sorts', sorts);
+        return tryInvoke(this, 'updateSorts', [sorts]);
+    }
+
+    /**
+     * Updates the visible columns based on `visibleColumns`
+     * (an array which includes columns based on their ability
+     * to fit within the table container)
+     *
+     * @memberof GavantTableComponent
+     */
+    @action
+    changeDisplay() {
```

Comment:
we should prob move this logic directly into the `updateColumnVisibility` method since its not really an action that is intended to be called anywhere else outside of that. and we might be able to get away with not having an extra loop on the columns aray/having to find them by valuePath, if we just do it directly in the updateColumnVisibility() (but be sure to comment all the affected lines when we move it there, so we know what to rip out once the real bug is fixed)